src/renderers/shared/fiber/ReactFiberStack.js

```diff
@@ -0,0 +1,101 @@
+/**
+ * Copyright 2013-present, Facebook, Inc.
+ * All rights reserved.
+ *
+ * This source code is licensed under the BSD-style license found in the
+ * LICENSE file in the root directory of this source tree. An additional grant
+ * of patent rights can be found in the PATENTS file in the same directory.
+ *
+ * @providesModule ReactFiberStack
+ * @flow
+ */
+
+'use strict';
+
+import type { Fiber } from 'ReactFiber';
+
+export type StackCursor<T> = {
+  current: T
+};
+
+const warning = require('warning');
+
+const valueStack : Array<any> = [];
+
+if (__DEV__) {
+  var fiberStack : Array<Fiber | null> = [];
+}
+
+let index = -1;
+
+exports.createCursor = function<T>(defaultValue : T) : StackCursor<T> {
+  return {
+    current: defaultValue,
+  };
+};
+
+exports.isEmpty = function() : boolean {
+  return index === -1;
+};
+
+exports.pop = function<T>(
+  cursor : StackCursor<T>,
+  fiber: Fiber,
+) : void {
+  if (index < 0) {
+    if (__DEV__) {
+      warning(false, 'Unexpected pop.');
+    }
+    return;
+  }
+
+  if (__DEV__) {
+    if (fiber !== fiberStack[index]) {
+      warning(false, 'Unexpected Fiber popped.');
+    }
+  }
+
+  cursor.current = index > 0
+    ? valueStack[index]
+    : (null : any);
```

Comment:
This is unsound since it doesn't reset to the defaultValue. However, I don't think this should ever be possible since the first value that goes into the value stack __is__ the defaultValue. So you shouldn't be able to pop further than that.